FIX Fix multiple severe bugs in non-metric MDS

dkobak opened the pull request:
#### Reference Issues/PRs
Fixes #16846.
Fixes #18933.  
Fixes #26999.
Fixes #27028.

#### What does this implement/fix? Explain your changes.

This fixes several bugs in non-metric MDS. The current implementation of non-metric MDS **has severe implementation errors and is broken.** See example in #27028: the loss increases after a couple of iterations (which should not happen) and algorithm stops. One error was pointed out in #18933: the upper-triangular part of the disparities matrix was not correctly copied into the lower-triangular part. I fixed it. The second error was that the isotonic regression should not be run before the first SMACOF iteration. The SMACOF update should happen first.

#26999 is a separate issue due to `IsotonicRegression` triggering `out_of_bounds` errors due to machine precision issues. Very easy fix via `out_of_bounds="clip"`.

#16846 is another separate issue: the returns stress value (MDS loss) does not correspond to the returned embedding, but to the previous iteration. Very easy fix by moving the stress computation after the SMACOF update.

For convenience, I also did a small API change: non-metric MDS uses `normalized_stress` and the current API for some reason does not allow `normalized_stress=True` when `metric=True`. I don't see any reason why this should not be allowed. It works just fine, and allows to compare normalized stress between metric and non-metric MDS. So I removed this restriction. It originated in #22562.

#### Any other comments?

Demonstration on the Digits dataset:

```Python
import pylab as plt
from sklearn.manifold import MDS
from sklearn.datasets import load_digits

X, y = load_digits(return_X_y=True)
mds = MDS(n_init=1, verbose=10, random_state=42, metric=False, normalized_stress=True, eps=1e-15)
Z = mds.fit_transform(X)
stress = mds.stress_

plt.figure(figsize=(4, 4))
plt.scatter(Z[:,0], Z[:,1], c=y, s=3, cmap="tab10")
plt.title(f"Digits dataset\nNEW non-metric MDS, norm. stress = {stress:.2f}")
```

Metric MDS:

![mds1](https://github.com/user-attachments/assets/5cb752fb-3b22-43ec-9265-9b66eb4392e6)

Non-metric MDS current:

![mds3](https://github.com/user-attachments/assets/0c2b82c7-82fa-4e2c-8ff6-a98d7691d6a3)

Non-metric MDS with this fix:

![mds2](https://github.com/user-attachments/assets/d8f316cb-6219-4910-9f27-a1511d93b433)

Apart from that, I would like to refactor the code of `_smacof_single()` for readability and also to do two API changes:
1. Current default is `n_init=4` which runs MDS four times. Other sklearn classes that offer this functionality use `n_init=1` by default, e.g. `sklearn.mixture.GaussianMixture`. This seems much more sensible to me, so I would suggest to switch to `n_init=1` here as well. The algorithm is slow as it is.
2. The convergence criterion is odd: it normalizes the stress by `np.sqrt((X**2).sum(axis=1)).sum()` but it is unclear why. The default `eps=1e-3` leads to bad underconvergence in some cases, e.g. on the Digits dataset in non-metric case. That's why I used `eps=1e-15` in my code above. I would like to change the convergence criterion to `(old_stress - stress) / stress < eps`.
3. I think the reasonable default initialization is not random but PCA. In case the input is not the data matrix but the distance matrix, one should replace PCA with PCoA (aka "classical MDS") and use that as initialization. Unfortunately, sklearn does not have classical MDS implemented; there are open PRs about that but they are stalled. If/when `ClassicalMDS` class is implemented, that should be use as the default initialization for non-metric (and also metric) MDS.

I did not include any of this into this PR because these are not "bug fixes" so I thought it should be separate PRs. But I can also see an argument for including (1) and (2) inside this PR as the changes are trivially simple. Let me know what you prefer.

dkobak committed:
Fix multiple bugs in nonmetric MDS

dkobak committed:
Add what's new

dkobak committed:
Remove obsolete test

dkobak committed:
Update example to pass doctest

adrinjalali commented:
@antoinebaker could you please have a look at this one?

antoinebaker commented:
Thanks a lot for the PR @dkobak ! It seems that there are indeed several implementation errors in `_smacof_single`.

As a general question, which reference do you follow to fix the implementation ? I have looked at two references so far:
1. Modern multidimensional scaling, Borg and Groenen. The algorithm is described in Smacof with Admissibly Transformed Proximities
2. smacof R package [article](https://www.jstatsoft.org/article/view/v031i03). The metric and nonmetric MDS are I think implemented in [smacofSym](https://github.com/cran/smacof/blob/master/R/smacofSym.R)
They slightly differ about this point:
> The second error was that the isotonic regression should not be run before the first SMACOF iteration. The SMACOF update should happen first.
I think 2. follows your idea, but 1. first do the monotonic regression first.

dkobak commented:
Hi @antoinebaker, thanks for looking into this. I would really like to have this fixed.

I think I looked at how this is done in R as the R implementation seems to me to be the reference one for smacof. It is co-developed by Jan de Leeuw who was the original author of the smacof algorithm (1977 paper). If I remember correctly, I looked at exactly the same paper as you found: https://www.jstatsoft.org/article/view/v102i10.

In general, it's an alternating algorithm, and as with any alternating algorithm it does not matter in principle which part to begin with. However, in practice, one way of starting it can be a better heuristic to ensure better convergence. I have not done many experiments, I just observed that on the Digits dataset, it only works one way. And as that is the way that is implemented in R, I thought we should follow that.

I suspect that with better initialization (via classical MDS, see my point (3) above) this issue of whether isotonic regression goes first or second should not matter. But classical MDS initialization would need to wait for another PR, so for now I would go ahead and mimic the R approach.

adrinjalali reviewed (commented):
Could you please add non-regression tests for all the issues this is fixing?

ogrisel commented on sklearn/manifold/_mds.py:
```diff
@@ -70,9 +70,8 @@ def _smacof_single(
+        Whether use and return normalized stress value (Stress-1) instead of raw
+        stress.
 
         .. versionadded:: 1.2
```
Please add a `.. versionchanged:: 1.7` entry to specify that `normalized_stress=True` is now also supported when `metric=False`.

dkobak replied:
Thanks, done.

ogrisel commented on sklearn/manifold/_mds.py:
```diff
@@ -84,7 +83,7 @@ def _smacof_single(
         The final value of the stress (sum of squared distance of the
         disparities and the distances for all constrained points).
-        If `normalized_stress=True`, and `metric=False` returns Stress-1.
+        If `normalized_stress=True`, returns Stress-1.
```
The following lines in the user guide also need to be updated accordingly:

https://github.com/scikit-learn/scikit-learn/blob/f4ba17bf88af25675c997400559f9e2e4c1cbe41/doc/modules/manifold.rst?plain=1#L479-L481

dkobak replied:
Thanks, done.

dkobak commented:
@ogrisel Hmm, I haven't seen this example before, but it seems that what happens on `main` is that the sign of the 1st axis of MDS/NMDS is flipped compared to X_true. If you were to flip the sign, then it would fit very well. So I think the fact that it "works better" on this branch is accidental. Somebody should just fix the sign of PC1 / PC2 in that example.

ogrisel commented:
> Could you please add non-regression tests for all the issues this is fixing?

In particular, please add non-regression tests for all the issues fixed by this PR as per-this PR's description. 

I checked that the reproducers of #26999 and #16846 are fixed with this PR. Those could probably be adapted as non-regression tests.

I am not sure how to turn #18933 into a non-regression test.

#27028 is nice, but too slow to run as a test. Maybe it could be turned into an MDS usage example, but we already have many examples in the example gallery and we are trying to reduce their number by consolidating them rather than add new ones.

ogrisel commented:
> I think I looked at how this is done in R as the R implementation seems to me to be the reference one for smacof

Maybe we could include a test on a toy dataset where scikit-learn used to produce bad results and check that we get the same results as computed by the R-based reference implementation, both for the stress values and the output of the `transform` method on a few arbitrary data points.

ogrisel reviewed (commented):
Some more inline feedback below:

dkobak committed:
Fix the MDS example

dkobak committed:
Fix the docs and docstrings regarding normalized_stress

dkobak commented:
> Some variable names in the examples are either confusing (clf for the PCA instance) or not very explicit (pos / npos). 

@ogrisel: I edited the example to fix the sign of PCs, and added it to this PR. Also did some minor edits to rename the variables and improve readability.

dkobak committed:
Add tests

dkobak committed:
Add one more test

dkobak commented:
I have added some tests @ogrisel. 

> I checked that the reproducers of https://github.com/scikit-learn/scikit-learn/issues/26999 and https://github.com/scikit-learn/scikit-learn/issues/16846 are fixed with this PR. Those could probably be adapted as non-regression tests.

Done, I added two tests for that.

> I am not sure how to turn https://github.com/scikit-learn/scikit-learn/issues/18933 into a non-regression test.

I am not sure either, and I think this is not needed.

> https://github.com/scikit-learn/scikit-learn/issues/27028 is nice, but too slow to run as a test.

I agree. We could add a test that running 3 iterations on Digits makes stress lower than running 2 iterations, but even that takes 2-3 seconds. Let me know if it's acceptable, but for now I did not include that.

> Maybe we could include a test on a toy dataset where scikit-learn used to produce bad results and check that we get the same results as computed by the R-based reference implementation

This makes sense but the problem is that I don't have a toy dataset on which NMDS failed before. I only observed it on Digits and other even larger datasets. On very small toy datasets (like in the usage example you showed above) NMDS worked fine.

At least I added a test that NMDS gives lower (normalized) stress than metric MDS. And that both can recover original data.

ogrisel commented on sklearn/manifold/tests/test_mds.py:
```diff
@@ -85,3 +111,37 @@ def test_normalized_stress_auto(metric, monkeypatch):
+
+def test_isotonic_outofbounds():
+    # This particular configuration can trigger out of bounds error
+    # in the isotonic regression (was fixed in PR 30514)
```
Rather than mention the PR which can be usually be found via git blame, we usually make it explicit that this is a non-regression test for a given issue number.

ogrisel commented on sklearn/manifold/tests/test_mds.py:
```diff
@@ -20,6 +20,43 @@ def test_smacof():
 
+def test_nonmetric_lower_normalized_stress():
+    # Testing that nonmetric MDS results in lower normalized stess compared
+    # compared to metric MDS
```
Please also link to the reference issue for this non-regression test.

ogrisel reviewed (approved):
I am not familiar with the MDS literature, so I trust you and @antoinebaker that this diff is actually in line with the reference implementation in R and the main references in the literature.

But from the look of the tests and the fact that this PR actually fixes the reproducers reported in the linked issues, this LGTM.

ogrisel commented:
> We could add a test that running 3 iterations on Digits makes stress lower than running 2 iterations, but even that takes 2-3 seconds. Let me know if it's acceptable, but for now I did not include that.

That feels a bit too slow. Let's hope the newly introduced tests are good enough as non-regression tests.

Once this is merged, +1 for follow-up PRs to address the other points raised in the descriptions.

dkobak committed:
Add issue number to test

dkobak committed:
Add random seed

dkobak commented on sklearn/manifold/tests/test_mds.py:
```diff
@@ -85,3 +111,37 @@ def test_normalized_stress_auto(metric, monkeypatch):
+    X = np.array([[1, 1], [1, 4], [1, 5], [3, 3]])
+    D = euclidean_distances(X)
+
+    mds_est = mds.MDS(n_components=2).fit(X)
```
Thanks, done.

dkobak committed:
Add refernce to the R implementation

dkobak committed:
Remove double line break

antoinebaker commented:
Thanks for all the work @dkobak.
I'll do a more thorough review soon, but regarding:

> > The convergence criterion is odd: it normalizes the stress by np.sqrt((X**2).sum(axis=1)).sum() but it is unclear why. The default eps=1e-3 leads to bad underconvergence in some cases, e.g. on the Digits dataset in non-metric case. That's why I used eps=1e-15 in my code above. I would like to change the convergence criterion to (old_stress - stress) / stress < eps.
> 
> Same thing / same question as above.

I think the R smacof code uses `old_stress - stress < eps` as a stopping criterion, where `stress` stands for the raw stress (maybe normalized by `nn=n_samples*(n_samples-1)/2` I have to check). As our goal is to follow the R implementation, we should use the same criterion. Actually a pretty good test is to check that we recover the R smacof results, so +1 for @ogrisel suggestion:

> > Maybe we could include a test on a toy dataset where scikit-learn used to produce bad results and check that we get the same results as computed by the R-based reference implementation
> 
> This makes sense but the problem is that I don't have a toy dataset on which NMDS failed before. I only observed it on Digits and other even larger datasets. On very small toy datasets (like in the usage example you showed above) NMDS worked fine.

For me that's okay that the small toy datasets don't break NMDS on main, as here the aim of the test would be to check the consistency with R smacof.

ogrisel commented:
I would rather do the api improvements with deprecation cycle in a follow PR to not delay the merge of the actual bugfixes.

ogrisel commented:
A single followup PR should be fine.

dkobak commented:
@antoinebaker Any changes to the stopping criterion would need to go through a deprecation cycle, and @ogrisel suggests to make those in a separate PR.

I don't think we necessarily need to mimic the R implementation _exactly_. I checked the stopping criterion there, and indeed it uses `(old_stress - stress) / nn < eps`, but the defaults are `eps=1e-6` and `max_iter=1000` (both different from our current defaults. I am okay with using the same stopping criterion, and can implement that in the follow-up PR.

antoinebaker commented on sklearn/manifold/_mds.py:
```diff
@@ -275,14 +289,18 @@ def smacof(
     normalized_stress : bool or "auto" default="auto"
-        Whether use and return normed stress value (Stress-1) instead of raw
-        stress calculated by default. Only supported in non-metric MDS.
+        Whether use and return normalized stress value (Stress-1) instead of raw
+        stress. By default, metric MDS uses raw stress while non-metric MDS uses
+        normalized stress.
```
```suggestion
        Whether to return normalized stress value (Stress-1) instead of raw
        stress. By default, metric MDS returns raw stress while non-metric MDS returns
        normalized stress.
```

dkobak replied:
Agree, removed "use", as you suggested.

dkobak commented on sklearn/manifold/_mds.py:
```diff
@@ -121,49 +120,59 @@ def _smacof_single(
+            disparities = np.zeros_like(dis_flat)
             disparities[sim_flat != 0] = disparities_flat
             disparities = disparities.reshape((n_samples, n_samples))
             disparities *= np.sqrt(
                 (n_samples * (n_samples - 1) / 2) / (disparities**2).sum()
             )
+            disparities = disparities + disparities.T
```
No, I think it's correct as written: the squared disparities should sum to `n * (n-1)`, or, equivalently, the upper triangle of the squared disparities should sum to `n * (n-1) / 2`.

antoinebaker replied:
You're right, that's the normalization mentioned in the paper and in the R code.

dkobak committed:
Rename some variables

dkobak commented on sklearn/manifold/_mds.py:
```diff
@@ -121,49 +120,59 @@ def _smacof_single(
 
-        dis = np.sqrt((X**2).sum(axis=1)).sum()
+        # Compute stress
+        dis = euclidean_distances(X)
```
Thanks, I agree. These are leftovers of the old code. I did the renaming.

dkobak commented on sklearn/manifold/_mds.py:
```diff
@@ -121,49 +120,59 @@ def _smacof_single(
+        normalization = np.sqrt((X**2).sum(axis=1)).sum()
         if verbose >= 2:
-            print("it: %d, stress %s" % (it, stress))
+            print(f"Iteration {it + 1:{len(str(max_iter))}}, stress {stress:.4f}")
```
Okay, if sklearn prefers less pretty verbose output but more readable code, then fine. Did the suggested edit.

dkobak commented on sklearn/manifold/_mds.py:
```diff
@@ -121,49 +120,59 @@ def _smacof_single(
+            print(f"Iteration {it + 1:{len(str(max_iter))}}, stress {stress:.4f}")
         if old_stress is not None:
-            if (old_stress - stress / dis) < eps:
+            if (old_stress - stress / normalization) < eps:
```
Yes, will do in the follow-up.

dkobak commented on sklearn/manifold/_mds.py:
```diff
@@ -275,8 +284,9 @@ def smacof(
-        stress calculated by default. Only supported in non-metric MDS.
+        Whether use and return normalized stress value (Stress-1) instead of raw
+        stress. By default, metric MDS uses raw stress while non-metric MDS uses
+        normalized stress.
```
Agree, removed "use", as you suggested.

dkobak commented on sklearn/manifold/tests/test_mds.py:
```diff
@@ -85,3 +143,39 @@ def test_normalized_stress_auto(metric, monkeypatch):
+    mds.smacof(dis, init=init, metric=False, n_init=1)
+
+
+def test_returned_stress():
+    # Test that the final stress corresponds to the final embedding
+    # (non-regression test for issue 16846)
+    X = np.array([[1, 1], [1, 4], [1, 5], [3, 3]])
+    D = euclidean_distances(X)
+
+    mds_est = mds.MDS(n_components=2, random_state=42).fit(X)
+    Z = mds_est.embedding_
+    stress = mds_est.stress_
+
+    D_mds = euclidean_distances(Z)
+    stress_Z = ((D_mds.ravel() - D.ravel()) ** 2).sum() / 2
+
+    assert_allclose(stress, stress_Z)
```
I suggest to do that in the follow-up PR, when we change the definition of normalized stress to normalize by distances in the embedding space (as you wrote in another comment).

antoinebaker replied:
I agree it makes more sense to test it when fixed in the follow up PR.

dkobak committed:
Fix the User Guide for MDS

dkobak commented on doc/modules/manifold.rst:
```diff
@@ -476,9 +476,7 @@ stress, is then defined by :math:`\sum_{i < j} d_{ij}(X) - \hat{d}_{ij}(X)`
-  The use of normalized Stress-1 can be enabled by setting `normalized_stress=True`,
-  however it is only compatible with the non-metric MDS problem and will be ignored
-  in the metric case.
+  The use of normalized Stress-1 can be enabled by setting `normalized_stress=True`.
```
Thank you. I agree that the User Guide was a mess. I have edited it, please take a look.

dkobak committed:
Fix the User Guide

dkobak commented:
@antoinebaker I'm not sure why the tests complain about the docs now, could you maybe help me with that?

antoinebaker commented on sklearn/manifold/_mds.py:
```diff
@@ -121,49 +120,59 @@ def _smacof_single(
+        # Compute stress
+        dis = euclidean_distances(X)
+        stress = ((dis.ravel() - disparities.ravel()) ** 2).sum() / 2
+        if normalized_stress:
+            stress = np.sqrt(stress / ((disparities.ravel() ** 2).sum() / 2))
```
Sounds good, I'll be happy to review the follow up PR, you can cc me when it will be ready for review.

dkobak committed:
fix the typo that has been fixed in upstream/main

dkobak committed:
Merge branch 'main' into non-metric-mds-bugfixes

antoinebaker reviewed (commented):
Thanks for User Guide, it's much better now. Not sure why the documentation fails on the CI, maybe try to merge main to see if the problem remains.

dkobak committed:
Update examples/manifold/plot_mds.py

Co-authored-by: antoinebaker <antoinebaker@users.noreply.github.com>

dkobak committed:
Update examples/manifold/plot_mds.py

Co-authored-by: antoinebaker <antoinebaker@users.noreply.github.com>

dkobak committed:
Update doc/modules/manifold.rst

Co-authored-by: antoinebaker <antoinebaker@users.noreply.github.com>

dkobak committed:
Update doc/modules/manifold.rst

Co-authored-by: antoinebaker <antoinebaker@users.noreply.github.com>

dkobak committed:
Update doc/modules/manifold.rst

Co-authored-by: antoinebaker <antoinebaker@users.noreply.github.com>

dkobak committed:
update docs

dkobak commented on doc/modules/manifold.rst:
```diff
@@ -418,67 +418,64 @@ Multi-dimensional Scaling (MDS)
 inequality), the distances between output two points are then set to be as
-close as possible to the similarity or dissimilarity data. In the non-metric
-version, the algorithms will try to preserve the order of the distances, and
+close as possible to the dissimilarity data. In the non-metric
```
Fair enough, edited.

dkobak committed:
update docs

dkobak commented:
> maybe try to merge main to see if the problem remains.

This did the trick.

I incorporated all your edits, and did some further edits to ensure consistent notation in the docs.

dkobak committed:
Fix an error in the example

antoinebaker commented on doc/modules/manifold.rst:
```diff
@@ -418,67 +418,64 @@ Multi-dimensional Scaling (MDS)
 
   .. math::
-      \sqrt{\frac{\sum_{i < j} (d_{ij} - \hat{d}_{ij})^2}{\sum_{i < j} d_{ij}^2}}.
+      \sqrt{\frac{\sum_{i < j} (\d_{ij} - \hat{d}_{ij})^2}{\sum_{i < j}
+      d_{ij}^2}}.
```
```suggestion
      \sqrt{\frac{\sum_{i < j} (\hat{d}_{ij} - d_{ij}(Z))^2}{\sum_{i < j}
      d_{ij}(Z)^2}}.
```

antoinebaker reviewed (commented):
A few typos.

dkobak committed:
Update doc/modules/manifold.rst

Co-authored-by: antoinebaker <antoinebaker@users.noreply.github.com>

dkobak committed:
Update doc/modules/manifold.rst

Co-authored-by: antoinebaker <antoinebaker@users.noreply.github.com>

dkobak committed:
Exclude verbose outputs from coverage tests

dkobak committed:
Merge branch 'non-metric-mds-bugfixes' of github.com:dkobak/scikit-learn into non-metric-mds-bugfixes

dkobak commented:
Thanks @antoinebaker! Now all checks have passed, yay! What's next?

antoinebaker commented on doc/modules/manifold.rst:
```diff
@@ -418,67 +418,64 @@ Multi-dimensional Scaling (MDS)
+  :math:`\delta_{ij} > \delta_{kl}`, then the embedding
+  seeks to enforce :math:`d_{ij}(Z) > d_{kl}(Z)`. A simple algorithm
+  to enforce proper ordination is to use an
+  isotonic regression of :math:`\delta_{ij}` on :math:`d_{ij}(Z)`, yielding
+  disparities :math:`\hat{d}_{ij}` on the same order as :math:`d_{ij}(Z)`.
```
`IsotonicRegression`of `y=distances` on `X=dissimilarities`, giving the `y_pred=disparities` close to the `distance` but ordered as the `dissimilarities`.
```suggestion
  isotonic regression of :math:`d_{ij}(Z)` on :math:`\delta_{ij}`, yielding
  disparities :math:`\hat{d}_{ij}` on the same order as :math:`\delta_{ij}`.
```

dkobak replied:
Hmm. No. It is correct that `IsotonicRegression` is run with `X=dissimilarities, y=distances`, but here `dissimilarities` are the input data to MDS, i.e. `\delta_ij`, and `distances` are distances in the embedding, i.e. `d_ij(Z)`. So I think it's correct to say that it is regressing "deltas on d(Z)" and that it yields disparities on the same order as "d(Z)".

dkobak replied:
Hmm, wait, I think I take that back.

dkobak replied:
Okay, sorry, I got confused about "regression on smth" terminology. Apparently, in regression we say that "y is regressed on X" (https://stats.stackexchange.com/questions/207425/why-do-we-say-the-outcome-variable-is-regressed-on-the-predictors). So, it should be "regression of d(Z) on deltas". Fine. I'm now trying to understand what's the correct formulation for the second part.

antoinebaker replied:
Oh that might be a grammar issue then ?
I would say: isotonic regression of `y` on `X`,  yielding estimate `y_pred` (close to `y`) ordered as `X`.
You would say the other way around ?

dkobak replied:
Okay, turns out I was confused about what "on the same order as" is supposed to mean :-) I thought it's about "order of magnitude" whereas you are clearly right -- it's about the "ordering"!

dkobak replied:
I merged your suggestion and did another reformulation to clarify that!

dkobak committed:
Update doc/modules/manifold.rst

Co-authored-by: antoinebaker <antoinebaker@users.noreply.github.com>

dkobak committed:
Clarify formulation

dkobak commented:
@adrinjalali I think this is ready. Take a look if you can approve it. @ogrisel already did.

After this is merged, I will start working on the follow-up PR concerning default `n_init` and the convergence criterion.

antoinebaker reviewed (approved):
LGTM, thanks @dkobak ! 
The CI failure is unrelated to the PR , see https://github.com/scikit-learn/scikit-learn/issues/31098.

dkobak committed:
Merge branch 'main' into non-metric-mds-bugfixes

adrinjalali commented on examples/manifold/plot_mds.py:
It would be really nice if in a separate PR, you could improve the structure of this example, to make it notebook style and add explanations to sections before each section.

adrinjalali reviewed (approved):
Awesome work!

dkobak commented:
Dear @ogrisel @adrinjalali @antoinebaker, I went ahead and created PR #31117 to change the defaults for `n_init` and the convergence criterion. Feedback very welcome over there. In particular, I wasn't sure how to approach the deprecation cycle, and if it's needed at all in this case. Cheers.
